### PR TITLE
Added native window handle wrapper for supported platforms

### PR DESCRIPTION
**Pull request by jayrulez**
Use NativeHandle in tests

There are a few things that I'm still considering.

1. The terminology for iOS is view rather than window but I kept it as WindowHandle rather than ViewHandle for consistency.

2. This abstraction can be used in the engine creation descriptor struct for OGL. However, since I am not able to test all platforms I am not comfortable making this change. 

I believe I need to forward declare HWND in Win32NativeWindow.h to use HWND instead of void* for the type. Including Windows.h there is problematic.

### CODE REVIEW

**Comment by DiligentGraphics** [comment]
Hi!
Linux build fails because some of the include files (xlib I think) `#define None` that creates a conflict.
Mac build apparently fails because cocoa headers can only be used when compiling objective c files.
Windows build fails because clang-format validator errors. Format validation is at this point only automated on Windows. If you generate files with `-D DILIGENT_BUILD_TESTS=TRUE`, there will be targets that automatically validate format, which also run as part of all build.

**Comment by DiligentGraphics** [comment]
Build problems on Linux and Mac is why I don't like to include system headers in public interface.
I suggest all system headers should be removed and forward declarations used if possible, or `void*`.

**Comment by jayrulez** [comment]
I updated the files to use void* and squashed all the fixup commits for a cleaner history.

**Comment by DiligentGraphics** [comment]
Thanks! I will merge this PR little later, have to do something else this week.

**Comment by jayrulez** [comment]
@DiligentGraphics 

Are you okay with changing 
```
    struct EngineGLCreateInfo : public EngineCreateInfo
    {
        /// Native window wrapper
        NativeWindow NativeWindow;
    };
```

to

```
    struct EngineGLCreateInfo : public EngineCreateInfo
    {
        /// Native window wrapper
        NativeWindow NativeWnd;
    };
```

 to fix the linux build?

or maybe rename the NativeWindow type to PlatformWindow?

**Comment by jayrulez** [comment]
I'm not very experienced with C++ so it was after the last commit that I realized I could have just used the SRO here:
```
struct EngineGLCreateInfo : public EngineCreateInfo
    {
        /// Native window wrapper
        ::NativeWindow NativeWindow;
    };
```

But I believe the last commit (the rename) is still okay.

File: Graphics/GraphicsEngine/interface/GraphicsTypes.h [hunk 1]
```diff
@@ -35,6 +35,7 @@
 #include "../../../Primitives/interface/BasicTypes.h"
 #include "../../../Primitives/interface/DebugOutput.h"
 #include "../../../Primitives/interface/FlagEnum.h"
+#include "../../../Platforms/interface/NativeWindowHandle.hpp"
```
Comment (DiligentGraphics) [inline Graphics/GraphicsEngine/interface/GraphicsTypes.h]: `.h` and `.hpp` extensions are used to clearly indicate which files can be used in both c and c++, and which are c++ only. Native window handle will be part of the core API and is compatible with both c and c++, so it should have .h extension

File: Platforms/Android/interface/AndroidNativeWindowHandle.hpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ *  out of the use or inability to use the software (including but not limited to damages 
+ *  for loss of goodwill, work stoppage, computer failure or malfunction, or any and 
+ *  all other commercial damages or losses), even if such Contributor has been advised 
+ *  of the possibility of such damages.
+ */
+
+#pragma once
+
+#include "../../Primitives/interface/CommonDefinitions.h"
+
```
Comment (DiligentGraphics) [inline Platforms/Android/interface/AndroidNativeWindowHandle.hpp]: Please enclose this in `DILIGENT_BEGIN_NAMESPACE`/`DILIGENT_END_NAMESPACE`

File: Platforms/Android/interface/AndroidNativeWindowHandle.hpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+struct AndroidNativeWindowHandle
+{
+    void* pWindowHandle DEFAULT_INITIALIZER(nullptr);
```
Comment (DiligentGraphics) [inline Platforms/Android/interface/AndroidNativeWindowHandle.hpp]: Since we are using void* for all handles, let's give them a more descriptive names: `pAWindow`

File: Platforms/Apple/interface/IOSNativeWindowHandle.hpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+#pragma once
+
+#include "../../Primitives/interface/CommonDefinitions.h"
+
+struct IOSNativeWindowHandle
+{
+    void* pWindowHandle DEFAULT_INITIALIZER(nullptr);
+};
```
Comment (DiligentGraphics) [inline Platforms/Apple/interface/IOSNativeWindowHandle.hpp]: `pNSView`

File: Platforms/Apple/interface/MacOSNativeWindowHandle.hpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ *  of the possibility of such damages.
+ */
+
+#pragma once
+
+#include "../../Primitives/interface/CommonDefinitions.h"
+
+struct MacOSNativeWindowHandle
+{
+    void* pWindowHandle DEFAULT_INITIALIZER(nullptr);
```
Comment (DiligentGraphics) [inline Platforms/Apple/interface/MacOSNativeWindowHandle.hpp]: `pNSView`

File: Platforms/Linux/interface/LinuxNativeWindowHandle.hpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+#pragma once
+
+#include "../../Primitives/interface/CommonDefinitions.h"
+
+struct LinuxNativeWindowHandle
+{
+    void* pDisplay      DEFAULT_INITIALIZER(nullptr);
+    void* pWindowHandle DEFAULT_INITIALIZER(nullptr);
```
Comment (DiligentGraphics) [inline Platforms/Linux/interface/LinuxNativeWindowHandle.hpp]: `pWindow`

File: Platforms/UWP/interface/UWPNativeWindowHandle.hpp [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+ *  of the possibility of such damages.
+ */
+
+#pragma once
+
+#include "../../Primitives/interface/CommonDefinitions.h"
+
+struct UWPNativeWindowHandle
+{
+    void* pWindowHandle DEFAULT_INITIALIZER(nullptr);
```
Comment (DiligentGraphics) [inline Platforms/UWP/interface/UWPNativeWindowHandle.hpp]: `pWindow`, I guess

File: Platforms/Win32/interface/Win32NativeWindowHandle.hpp [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+ *  of the possibility of such damages.
+ */
+
+#pragma once
+
+#include "../../Primitives/interface/CommonDefinitions.h"
+
+struct Win32NativeWindowHandle
+{
+    void* pWindowHandle DEFAULT_INITIALIZER(nullptr);
```
Comment (DiligentGraphics) [inline Platforms/Win32/interface/Win32NativeWindowHandle.hpp]: `hWnd`

File: README.md [hunk 9]
```diff
@@ -205,9 +205,9 @@ attaches to the one initialized by the app. An example of the engine initializat
 
 On MacOS, Diligent Engine supports OpenGL and Vulkan backends. Initialization of GL context on MacOS is
 performed by the application, and the engine attaches to the context created by the app; see
-[GLView.m](https://github.com/DiligentGraphics/DiligentEngine/blob/master/Common/NativeApp/Apple/Source/Classes/OSX/GLView.m)
+[GLView.m](https://github.com/DiligentGraphics/DiligentTools/tree/master/NativeApp/Apple/Source/Classes/OSX/GLView.m)
```
Comment (DiligentGraphics) [inline README.md]: Thanks, did not know these were broken. Could you submit this a separate PR?

File: Graphics/GraphicsEngine/interface/GraphicsTypes.h [hunk 10]
```diff
@@ -1318,21 +1319,11 @@ struct EngineCreateInfo
...
-    /// * On Android platform, this is a pointer to the native window (ANativeWindow*)
-    /// * On Linux, this is the native window (Window)
-    void* pNativeWndHandle DEFAULT_INITIALIZER(nullptr);
-
-#if PLATFORM_LINUX
-    /// For linux platform only, this is the pointer to the display
-    void* pDisplay         DEFAULT_INITIALIZER(nullptr);
-#endif
+	/// Native window wrapper
+	NativeWindowHandle NativeWindow;
```
Comment (DiligentGraphics) [inline Graphics/GraphicsEngine/interface/GraphicsTypes.h]: Let's just call it `NativeWindow Window`

File: Platforms/Apple/interface/MacOSNativeWindowHandle.hpp [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ *  or consequential damages of any character arising as a result of this License or 
+ *  out of the use or inability to use the software (including but not limited to damages 
+ *  for loss of goodwill, work stoppage, computer failure or malfunction, or any and 
+ *  all other commercial damages or losses), even if such Contributor has been advised 
+ *  of the possibility of such damages.
+ */
+
+#pragma once
+
+#include "../../Primitives/interface/CommonDefinitions.h"
```
Comment (DiligentGraphics) [inline Platforms/Apple/interface/MacOSNativeWindowHandle.hpp]: Should be `../../../` This is why Linux/Mac build fails.

File: Platforms/Apple/interface/MacOSNativeWindowHandle.hpp (hunk 5 above)
Comment (jayrulez) [reply Platforms/Apple/interface/MacOSNativeWindowHandle.hpp]: MacOS uses a NSWindow so should we not just use pWindow here? or pNSWindow to be fore specific?

File: Graphics/GraphicsEngine/interface/GraphicsTypes.h (hunk 10 above)
Comment (jayrulez) [reply Graphics/GraphicsEngine/interface/GraphicsTypes.h]: Should I rename the NativeWindowHandle.h file to NativeWindow.h?

File: README.md (hunk 9 above)
Comment (jayrulez) [reply README.md]: Yes, I can submit these fix in another PR. I believe there were other broken links too. I will try to find and fix them all.

File: Platforms/interface/NativeWindowHandle.h [hunk 12]
```diff
@@ -32,33 +32,63 @@
 #if PLATFORM_WIN32
 
 #    include "../Win32/interface/Win32NativeWindowHandle.hpp"
+
+DILIGENT_BEGIN_NAMESPACE(Diligent)
+
 typedef struct Win32NativeWindowHandle NativeWindowHandle;
 
+DILIGENT_END_NAMESPACE // namespace Diligent
```
Comment (jayrulez) [inline Platforms/interface/NativeWindowHandle.h]: Ideally, I want to get rid of these duplicate namespace declaration in this file. Maybe I should just move the typedef to the platform specific file and just include it here?

Comment (DiligentGraphics) [reply Platforms/interface/NativeWindowHandle.h]: Ah, OK. Note that the rest of the platform-specific classes not being in a namespace is a mistake which I am planning to fix at some point.
I would structure the file as follows (and then fix the other headers similar later):

```cpp

#if PLATFORM_WIN32
#    include "../Android/interface/Win32NativeWindow.h"
...
#endif

DILIGENT_BEGIN_NAMESPACE(Diligent)

#if PLATFORM_WIN32
typedef struct Win32NativeWindow NativeWindow;
...
#endif

DILIGENT_END_NAMESPACE
```

There will be duplicated `#if-#elseif` struture, but this is probably a better way then having  multiple `BEGIN_NAMESPACE`/`END_NAMESPACE` sections

File: Platforms/Apple/interface/MacOSNativeWindowHandle.hpp (hunk 5 above)
Comment (DiligentGraphics) [reply Platforms/Apple/interface/MacOSNativeWindowHandle.hpp]: +1 for `pNSWindow`

File: Graphics/GraphicsEngine/interface/GraphicsTypes.h (hunk 10 above)
Comment (DiligentGraphics) [reply Graphics/GraphicsEngine/interface/GraphicsTypes.h]: Yes please

**Comment by jayrulez** [comment]
@DiligentGraphics if everything looks good then I'll squash into 1 commit for a cleaner history but I want your approval first as I believe squashing will cause us to lose the review history.

File: Tests/DiligentCoreAPITest/include/TestingEnvironment.hpp [hunk 13]
```diff
@@ -12,7 +12,7 @@
  *  distributed under the License is distributed on an "AS IS" BASIS,
  *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  *  See the License for the specific language governing permissions and
- *  limitations under the License.
+ *  limitations under the License.hWnd
```
Comment (DiligentGraphics) [inline Tests/DiligentCoreAPITest/include/TestingEnvironment.hpp]: Typo has sneaked

**Comment by DiligentGraphics** [comment]
Looks good to me!

File: ReleaseHistory.md [hunk 14]
```diff
@@ -8,7 +8,7 @@
 
 ### API Changes
 
-* Added `NativeWindowHandle` wrapper and replaced `pNativeWndHandle` and `pDisplay` members with it in `EngineGLCreateInfo` (API Version 2400XX)
+* Added `NativeWindow` wrapper and replaced `pNativeWndHandle` and `pDisplay` members with it in `EngineGLCreateInfo` (API Version 2400XX)
```
Comment (DiligentGraphics) [inline ReleaseHistory.md]: Please update this to 240053, as well as [here](https://github.com/DiligentGraphics/DiligentCore/blob/master/Graphics/GraphicsEngine/interface/APIInfo.h#L33). I also updated the file with C interface that I forgot to mentiond.

**Comment by jayrulez** [comment]
Should be ready now. Thanks for the review and patience.

**Comment by DiligentGraphics** [comment]
This PR is merged now: https://github.com/DiligentGraphics/DiligentCore/commit/dadb44ef46cb83a957507a1d2d5d20986b3cec22
I made few changes which confused GitHub, so I the PR did not close automatically.
Thank you for help!
